feat(trace-view): Improve how we call eventstore.get_events

wmak opened the pull request:
- This allows eventstore.get_events to directly take a list of events
  to retrieve, currently get_events takes a filter of project and
  event ids.
  - This is problematic for the trace view where we already know the
    project to event_id mapping, and so we're not getting the
    optimization mentioned in the get_events function.

wmak committed:
feat(trace-view): Improve how we call eventstore.get_events

- This allows eventstore.get_events to directly take a list of events
  to retrieve, currently get_events takes a filter of project and
  event ids.
  - This is problematic for the trace view where we already know the
    project to event_id mapping, and so we're not getting the
    optimization mentioned in the get_events function.

wmak commented on src/sentry/eventstore/snuba/backend.py:
```diff
@@ -92,26 +94,34 @@ def __get_events(
             and filter.project_ids
-            and len(filter.event_ids) * len(filter.project_ids) < min(limit, NODESTORE_LIMIT)
+            and (
+                len(filter.event_ids) * len(filter.project_ids) < min(limit, NODESTORE_LIMIT)
```
so with trace-view, we can get up to 100 event_ids, and a large number of project_ids, but we'll ever only want to get up to 100 events, staying below this limit.
But since this is creating a list of events for every project + event we'll usually go over this limit and miss this optimization.

wmak commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -168,9 +169,14 @@ def serialize(self, parent_map, root, warning_extra, params, snuba_event=None, e
                 event.event_id: event
                 for event in eventstore.get_events(
                     eventstore.Filter(
+                        start=params["start"],
+                        end=params["end"],
```
Unrelated fix, this corrects an issue where we're doing a 90d query in snuba for these events.

lynnagara commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -168,9 +169,14 @@ def serialize(self, parent_map, root, warning_extra, params, snuba_event=None, e
-                    )
+                    ),
+                    event_list=[
+                        Event(project_id=event["project.id"], event_id=event["id"])
+                        for event in parent_map.values()
```
The eventstore should create the actual event instances, like it does in the rest of this method. Doing it this way only if an event list is specified is rather inconsistent. I think it'd be better to just pass a list of project id / event id tuples.

lynnagara commented on src/sentry/eventstore/base.py:
```diff
@@ -146,6 +147,7 @@ def get_events(
         limit (int): Query limit - default 100
         offset (int): Query offset - default 0
         referrer (string): Referrer - default "eventstore.get_events"
+        event_list (Sequence[Event]): List of events to get
```
project_id / event_id filtering is the role of the `snuba_filter`. Can we add this argument to the `snuba_filter` instead?

markstory commented on tests/sentry/eventstore/snuba/test_backend.py:
```diff
@@ -94,6 +95,27 @@ def test_get_events(self):
+                Event(
+                    project_id=self.project2.id,
+                    event_id="b" * 32,
+                ),
```
Seems curious to be giving a list of events to eventstore to get the events. Could there be another structure we use instead? Something like an `EventReference` or `EventIdentifier`?

wmak replied:
Hmm good point :+1: I'll create `EventIdentifier`, I think get_events can return stubbed events which might get confused with `EventReference`

wmak committed:
ref: Moving from event_list to eventIdentifiers

- Makes this more consistent with the existing api where we use the
  filter instead

lynnagara commented on src/sentry/eventstore/models.py:
```diff
@@ -29,6 +29,21 @@ def ref_func(x):
+        self.project_id = project_id
+        self.event_id = event_id
+
+    def to_event(self):
+        return Event(self.project_id, self.event_id)
```
Please don't introduce a dependency of the Snuba filter to the Snuba event model (filter shouldn't know anything about the model). Eventstore should be the one creating the Event objects like it for all of the rest of the methods. I'd suggest making EventIdentifier a tuple (or a named tuple if you prefer) and move it out of the models.py file.

wmak commented:
Closing this in favour of #24160, see PR description there for explanation of why
